src/lsm/tree.zig

```diff
@@ -658,26 +658,28 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type) type {
             return if (value == null or tombstone(value.?)) null else value.?;
         }
 
-        pub fn open(tree: *Tree, callback: *const fn (*Tree) void) void {
-            assert(tree.open_callback == null);
-            tree.open_callback = callback;
-            tree.compaction_op = tree.grid.superblock.working.vsr_state.commit_min;
+        pub fn open_commence(tree: *Tree, manifest_log: *ManifestLog) void {
+            assert(tree.compaction_op == null);
+            assert(tree.key_range == null);
 
-            tree.manifest.open(manifest_open_callback);
+            tree.manifest.open_commence(manifest_log);
         }
 
-        fn manifest_open_callback(manifest: *Manifest) void {
-            const tree = @fieldParentPtr(Tree, "manifest", manifest);
-            assert(tree.open_callback != null);
+        pub fn open_table(tree: *Tree, level: u8, table: *const schema.ManifestLog.TableInfo) void {
+            assert(tree.compaction_op == null);
+            assert(tree.key_range == null);
 
-            const callback = tree.open_callback.?;
-            tree.open_callback = null;
+            const tree_table = Manifest.TreeTableInfo.decode(table);
+            tree.manifest.levels[level].insert_table(tree.manifest.node_pool, &tree_table);
+        }
 
+        pub fn open_complete(tree: *Tree) void {
+            assert(tree.compaction_op == null);
             assert(tree.key_range == null);
-            tree.key_range = manifest.key_range();
-            maybe(tree.key_range == null);
 
-            callback(tree);
+            tree.compaction_op = tree.grid.superblock.working.vsr_state.commit_min;
```

Comment:
Let's assert that `tree.compaction_op` points at the bar boundary. 

And let's assert here that the manifest is consistent (no interior empty levels, key ranges are disjoint for the latest snapshot, etc).

Reply:
:thinking: What do you mean by "interior empty levels"?

Reply:
I think that if level i is not empty, then level i-1 can’t be empty. That is, empty levels are always a suffix of the atray

Reply:
Ah, I see! I think we can only safely check this for `TableUsage.general` tables (object and id trees):

```zig
            switch (Table.usage) {
                // Interior levels are non-empty.
                .general => {
                    var empty: bool = false;
                    for (&manifest.levels) |*level| {
                        var level_iterator =
                            level.iterator(.visible, &.{snapshot}, .ascending, null);
                        if (level_iterator.next()) |_| {
                            assert(!empty);
                        } else {
                            empty = true;
                        }
                    }
                },
                // In the secondary index TableUsage, it is possible (albeit unlikely!) that every
                // table in an interior level is deleted.
                //
                // Unlike general-usage tables, secondary-index tombstones need not compact down to
                // the last level of the tree before they are deleted. (Rather, the tombstones are
                // deleted as soon as they merge with their corresponding "put").
                // In this way, enough object deletions may lead to compactions where the both input
                // tables entirely cancel each other out, and no output table is written at all.
                // See `TableUsage` for more detail.
                .secondary_index => {},
            }
```